clients/src/main/java/org/apache/kafka/common/record/AbstractLegacyRecordBatch.java

```diff
@@ -264,26 +263,48 @@ static void writeHeader(DataOutputStream out, long offset, int size) throws IOEx
     }
 
     private static final class DataLogInputStream implements LogInputStream<AbstractLegacyRecordBatch> {
-        private final DataInputStream stream;
+
+        private static final int OFFSET_OFFSET = 0;
+        private static final int SIZE_OFFSET = 8;
+        private static final int BUFFER_SIZE = 8 // offset
+                                               + 4; // size
+        private final InputStream stream;
         protected final int maxMessageSize;
+        private final ByteBuffer headerBuffer;
 
-        DataLogInputStream(DataInputStream stream, int maxMessageSize) {
+        DataLogInputStream(InputStream stream, int maxMessageSize) {
             this.stream = stream;
             this.maxMessageSize = maxMessageSize;
+            this.headerBuffer = ByteBuffer.allocate(BUFFER_SIZE);
         }
 
         public AbstractLegacyRecordBatch nextBatch() throws IOException {
             try {
-                long offset = stream.readLong();
-                int size = stream.readInt();
+                int n = 0;
+                while (n < BUFFER_SIZE) {
```

Comment:
You and I had the exact same thought: https://github.com/apache/kafka/pull/3164/commits/a7bc3e6e98ad3f983f5619601b216e83142afe42